src/rtcSessionHelpers.ts

```diff
@@ -15,29 +15,92 @@ limitations under the License.
 */
 
 import { MatrixRTCSession } from "matrix-js-sdk/src/matrixrtc/MatrixRTCSession";
+import { AutoDiscovery, RoomMember } from "matrix-js-sdk";
+import { logger } from "matrix-js-sdk/src/logger";
+import { Focus } from "matrix-js-sdk/src/matrixrtc/focus";
+import {
+  LivekitFocus,
+  LivekitFocusActive,
+  LivekitFocusConfig,
+  isLivekitFocus,
+  isLivekitFocusConfig,
+} from "matrix-js-sdk/src/matrixrtc/LivekitFocus";
 
 import { PosthogAnalytics } from "./analytics/PosthogAnalytics";
-import { LivekitFocus } from "./livekit/LivekitFocus";
 import { Config } from "./config/Config";
 import { ElementWidgetActions, WidgetHelpers, widget } from "./widget";
 
-function makeFocus(livekitAlias: string): LivekitFocus {
-  const urlFromConf = Config.get().livekit!.livekit_service_url;
-  if (!urlFromConf) {
-    throw new Error("No livekit_service_url is configured!");
-  }
+const FOCI_WK_KEY = "org.matrix.rtc_foci";
 
+export function makeActiveFocus(): LivekitFocusActive {
   return {
     type: "livekit",
-    livekit_service_url: urlFromConf,
-    livekit_alias: livekitAlias,
+    focus_selection: "oldest_membership",
   };
 }
 
-export function enterRTCSession(
+async function makePreferredFoci(
+  rtcSession: MatrixRTCSession,
+  livekitAlias: string,
+): Promise<LivekitFocus[]> {
+  logger.log("Start building foci_preferred list: ", rtcSession.room.roomId);
+
+  const preferredFoci: LivekitFocus[] = [];
+
+  // Make the foci from the running rtc session the highest priority one
+  // This minimizes how often we need to switch foci during a call.
+  const focusFromMatrixRTC = rtcSession
+    .getOldestMembership()
+    ?.getPreferredFoci()[0];
+  if (focusFromMatrixRTC && isLivekitFocus(focusFromMatrixRTC)) {
+    logger.log("Adding livekit focus from oldest member: ", focusFromMatrixRTC);
+    preferredFoci.push(focusFromMatrixRTC);
+  }
+
+  // Prioritize the client well known over the configured sfu.
+  const wellKnownFoci = rtcSession.room.client.getClientWellKnown()?.[
+    FOCI_WK_KEY
+  ] as Focus[];
+  if (wellKnownFoci.length > 0) {
+    logger.log("Adding livekit focus from well known: ", wellKnownFoci);
+    preferredFoci.push(
+      ...wellKnownFoci
+        .filter(isLivekitFocusConfig)
```

Comment:
Because `isLivekitFocusConfig` (and the others) don't handle `null`/`undefined`, maybe this is a good idea:
```suggestion
        .filter(f => !!f)
        .filter(isLivekitFocusConfig)
```